pkg/diagnose/check.go

```diff
@@ -24,10 +26,92 @@ func init() {
 	diagnosis.Register("check-datadog", diagnose)
 }
 
+func diagnose(diagCfg diagnosis.Config) []diagnosis.Diagnosis {
+	if diagCfg.RunningInAgentProcess {
+		return diagnoseInAgentProcess(diagCfg)
+	}
+
+	return diagnoseInCLIProcess(diagCfg)
+}
+
+func partitionChecks(checks []check.Check) [][]check.Check {
+	instancesMap := map[string][]check.Check{}
+
+	// partition per check (instances of a check groupped by check/integration)
+	for _, c := range checks {
+		checkName := c.String()
+		instances, exists := instancesMap[checkName]
+		if exists {
+			instances = append(instances, c)
+		} else {
+			instances = make([]check.Check, 1)
+			instances[0] = c
+		}
+		instancesMap[checkName] = instances
+	}
+
+	// copy map to array
+	instancesArr := make([][]check.Check, 0, len(instancesMap))
+	for _, instances := range instancesMap {
+		instancesArr = append(instancesArr, instances)
+	}
+
+	// sort this array
+	sort.Slice(instancesArr, func(i, j int) bool {
+		return instancesArr[i][0].String() < instancesArr[j][0].String()
+	})
+
+	return instancesArr
+}
+
+func getInstanceDiagnoses(instance check.Check) []diagnosis.Diagnosis {
+
+	// Get diagnoses
+	diagnoses, err := instance.GetDiagnoses()
+	if err == nil && len(diagnoses) > 0 {
+		// Set category as check name if it was not set
+		for i, d := range diagnoses {
+			if len(d.Category) == 0 {
+				diagnoses[i].Category = instance.String()
+			}
+		}
+	} else if err != nil {
+		// Check Diagnose method return error
+		diagnoses = append(diagnoses, diagnosis.Diagnosis{
+			Result:    diagnosis.DiagnosisUnexpectedError,
+			Name:      string(instance.ID()),
+			Category:  instance.String(),
+			Diagnosis: "Check Dianose failes with unexpected errors",
+			RawError:  err,
+		})
+	}
+
+	return diagnoses
+}
+
+func diagnoseInAgentProcess(diagCfg diagnosis.Config) []diagnosis.Diagnosis {
+	// get checks
+	checks := common.Coll.GetChecks()
+
+	partInstances := partitionChecks(checks)
```

Comment:
Is there any reason to run diagnostics in any particular order?

The whole partitionChecks and two-level slice can probably be replaced by sorting the checks slice by check ID (which includes the integration name).

Reply:
Yes, I need to group the by category, which in this case it would be check (not its instance name) by default. There are some other potential benefits of two level array (tree conceptually) but currently they are not coded. For now I have rewrite per your suggestion completely eliminating additional function.
Thank you.

Reply:
Primararly to keep diagnoses grouped by category because it would be strange in my opinion report two related diagnoses far apart. It is also would be strange in my opinion again to have two instances of a check interspersed with diagnoses from other checks.

Reply:
Yep, makes sense. So if this is a presentation concern, I think it would be best to handle at the presentation time. Would it be possible to sort `[]Diagnosis` before printing it instead? It would be even more clear, because then you can sort by `Category` explicitly.

Reply:
Theoretically yes but I still prefer to return diagnoses from "backend" already organized as well as query them. I do not think it is principal and I also feel that it may need to be changed in future by I prefer for now go this way (call it gut feeling or my internal "esthetics").

Reply:
Then I guess you also need to sort checks in the other codepath that generates diagnoses: https://github.com/DataDog/datadog-agent/pull/18792/files#diff-c792372d6a3476ee5cf3099091bfd5ce07d3f7223bd4d4627a75f88444a380b8R152-R154

Reply:
On one hand, in this case, they are "naturally" grouped by check (there is a double nested loop), but on other hand, the visual order of instances coming when checks are run local vs Agent would be different and that is not good thing, not only in this cases but potentially for others. I will follow your advise and sort before output to make list of output in consistent order no matter what the method. Thank you